### PR TITLE
SPIR-V & some minor fixes

### PR DESCRIPTION
**Pull request by franz**
Adds a SPIR-V "mode" of OpenCL CTS (through offline compilation), plus a bunch of minor fixes.

Tested with CPU extensively, with TCE just lightly, with HSA / CUDA / Vulkan not at all.

### CODE REVIEW

**Review by pjaaskel — APPROVED** [review]
Great work! Can you add documentation for how to enable the SPIR-V mode in the conformance tester? I'll test this PR shortly...

File: cl_offline_compiler.sh.in.cmake [hunk 1]
```diff
@@ -0,0 +1,208 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (franz) [reply cl_offline_compiler.sh.in.cmake:1]: ok

File: lib/kernel/SPIR/generate_spir_wrapper.py [hunk 2]
```diff
@@ -1300,15 +1305,15 @@ def gen_three_variants(f, ret_type, ret_ext, AS, args, orders):
...
+attributes #0 = { alwaysinline nounwind "correctly-rounded-divide-sqrt-fp-math"="false" "denorms-are-zero"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protector-buffer-size"="8" "unsafe-fp-math"="false" "use-soft-float"="false" }
 
 !opencl.ocl.version = !{!0}
 !opencl.spir.version = !{!0}
 !llvm.ident = !{!1}
 !llvm.module.flags = !{!2, !3}
 
 !0 = !{i32 1, i32 2}
-!1 = !{!"clang version 6.0.0"}
+!1 = !{!"clang version 12.0.0"}
```
[the remark this answers is not in the document]
Comment (franz) [reply lib/kernel/SPIR/generate_spir_wrapper.py:1343]: good point. I'm not sure if any tool requires this MD. maybe we can drop and see if anything breaks.

**Comment by franz** [comment]
Hmm ok, tested with Vulkan no errors there, tested with CUDA, there one test fails with a very weird error "CUDA_ERROR_LAUNCH_FAILED: unspecified launch failure".

File: cl_offline_compiler.sh.in.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (franz) [reply cl_offline_compiler.sh.in.cmake:1]: fixed

**Comment by pjaaskel** [comment]
Can you rebase now? Also is it still not documented how to enable testing the conformance suite via SPIR-V builds?

**Comment by franz** [comment]
> Can you rebase now?

done

> still not documented how to enable testing the conformance suite via SPIR-V builds?

the CTS test labels with _spirv are automatically available if you build pocl with SPIRV support.

**Comment by pjaaskel** [comment]
For CUDA (LLVM 14) here: 147 - examples/example2_spir (Failed)

**Comment by franz** [comment]
CUDA should be fixed now

**Pull request merged**